### PR TITLE
FileGlobs.py: Recursively look for gitignore file

### PR DESCRIPTION
**Pull request by Wert1996**
This enables us to look for .gitignore files
recursively in the subdirectories starting
from root.

Fixes https://github.com/coala/coala-quickstart/issues/108

### CODE REVIEW

File: coala_quickstart/generation/FileGlobs.py [hunk 1]
```diff
@@ -24,7 +24,12 @@ def get_project_files(log_printer, printer, project_dir, non_interactive=False):
     file_globs = ["**"]
 
     ignore_globs = None
-    if os.path.isfile(os.path.join(project_dir, ".gitignore")):
+    flag_gitignore = 0
+    for dirName, subdirName, fileList in os.walk(project_dir):
```
Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: Please use snake_case for consistency as it is used everywhere in the project :)

File: coala_quickstart/generation/FileGlobs.py [hunk 2]
```diff
@@ -24,7 +24,12 @@ def get_project_files(log_printer, printer, project_dir, non_interactive=False):
...
+        if os.path.isfile(os.path.join(dirName, ".gitignore")):
+            flag_gitignore = 1
```
Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: Instead of storing a flag, may be directly store the path of gitignore, so you don't have to traverse again at https://github.com/coala/coala-quickstart/pull/110/files#diff-ded5849e720581d4ad8ada46114a8731R71

Comment (jayvdb) [inline coala_quickstart/generation/FileGlobs.py]: What if there are many `.gitignore` , shouldn't they all be imported?

Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: Yeah, just found out that git allows multiple `.gitignore` , so we must collect all of the paths in a list

File: coala_quickstart/generation/Utilities.py [hunk 3]
```diff
@@ -68,12 +68,13 @@ def get_gitignore_glob(project_dir, filename=".gitignore"):
         A list generator of glob expressions generated from the
         ``.gitignore`` file.
     """
-    gitignore = os.path.join(project_dir, filename)
-
-    with open(gitignore, "r") as file:
-        for line in file:
-            for glob in parse_gitignore_line(line):
-                yield os.path.join(project_dir, glob)
+    for dirName, subdirList, fileList in os.walk(project_dir):
```
Comment (satwikkansal) [inline coala_quickstart/generation/Utilities.py]: Not required if you collect all the paths at https://github.com/coala/coala-quickstart/pull/110/files#diff-7e6eaad4d79d2cfe0c15b00a0c204498R29 and pass them as an argument.

File: coala_quickstart/generation/FileGlobs.py (hunk 2 above)
Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: @Wert1996 Another thing to consider is if the globs specified in `.gitignore` are relative to the subdirectory in which the `.gitignore` is present or not

File: coala_quickstart/generation/FileGlobs.py [hunk 4]
```diff
@@ -24,7 +24,12 @@ def get_project_files(log_printer, printer, project_dir, non_interactive=False):
     file_globs = ["**"]
```
Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: Your commit contains couple of extra files which are not really required, please remove them :sweat_smile:

File: tests/generation/FileGlobs.py [hunk 5]
```diff
@@ -71,12 +72,22 @@ def test_get_project_files_gitignore(self):
...
         files += [".gitignore"]
+        os.chdir("another_folder")
+        with open('.gitignore', 'w') as f:
+            f.write("""
+# Start of gitignore
+*.js
+# End of gitignore""")
+        os.chdir(os.path.dirname(os.path.realpath(__file__)))
+        os.chdir("file_globs_gitignore_testfiles")
+        files += [os.path.join("another_folder", ".gitignore")]
```
Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: Add a test case for multiple `.gitignore` files also :)

File: coala_quickstart/generation/FileGlobs.py (hunk 2 above)
Comment (Wert1996) [inline coala_quickstart/generation/FileGlobs.py]: @satwikkansal .gitignore IS relative to the subdirectory. Confirmed it by trying.

Comment (satwikkansal) [inline coala_quickstart/generation/FileGlobs.py]: oh, cool :)

File: tests/generation/FileGlobs.py (hunk 5 above)
Comment (Wert1996) [inline tests/generation/FileGlobs.py]: There are already 2 .gitignore files in the test. Should I add any other than those?

Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: Not really required, but having 3 gitignores in different levels and then aggregating ignore globs from them would be more robust test case. Anyways, it's up to you

File: coala_quickstart/generation/Utilities.py [hunk 6]
```diff
@@ -57,7 +57,7 @@ def parse_gitignore_line(line):
             yield line
 
 
-def get_gitignore_glob(project_dir, filename=".gitignore"):
+def get_gitignore_glob(project_dir, gitignore_dir_list, filename=".gitignore"):
```
Comment (satwikkansal) [inline coala_quickstart/generation/Utilities.py]: Please describe the `gitignore_dir_list` in the inline documentation.

File: tests/generation/FileGlobs.py [hunk 7]
```diff
@@ -71,14 +74,39 @@ def test_get_project_files_gitignore(self):
...
+            os.path.join("another_folder", "script.js"),
+            os.path.join("data", "sample", "index.html"),
             "run.pyc"]
 
         for file in files + ignored_files:
             os.makedirs(os.path.dirname(os.path.abspath(file)), exist_ok=True)
             open(file, "w").close()
         files += [".gitignore"]
+        os.chdir("another_folder")
+        with open('.gitignore', 'w') as f:
```
Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: It would be slightly more intuitive to read up the code if we write these two new  `.gitignore` files just after the first `.gitignore` because you're adding `scripts.js` and `index.html` to ignored files before defining the relevant gitignores.

**Comment by satwikkansal** [comment]
Also needs rebase

File: tests/generation/FileGlobs.py [hunk 8]
```diff
@@ -56,14 +56,34 @@ def test_get_project_files_gitignore(self):
...
+        os.chdir("file_globs_gitignore_testfiles")
         files = [os.path.join("src", "main.c"),
             os.path.join("src", "main.h"),
             os.path.join("src", "lib", "ssl.c"),
             os.path.join("src", "tests", "main.c"),
             os.path.join("src", "main.py"),
             os.path.join("src", "upload.c"),
             os.path.join("another_folder", "another_file.c"),
-            os.path.join("data", "sample","index.css"),
+            os.path.join("data", "sample", "index.css"),
```
Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: Irrelevant change, we have an issue for this https://github.com/coala/coala-quickstart/issues/113 , so if you want to do this change, please do all of the changes in separate commit, or better, separate PR :)

File: tests/generation/FileGlobs.py [hunk 9]
```diff
@@ -82,30 +102,12 @@ def test_get_project_files_gitignore(self):
...
-        os.chdir("sample")
-        with open('.gitignore', 'w') as f:
-            f.write("""
-# Start of gitignore
-*.html
-# End of gitignore""")
-        os.chdir(os.path.dirname(os.path.realpath(__file__)))
-        os.chdir("file_globs_gitignore_testfiles")
-        files += [os.path.join("data","sample",".gitignore")]
+        files += [os.path.join("data", "sample", ".gitignore")]
```
Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: Same as https://github.com/coala/coala-quickstart/pull/110/files/749fc0d37f4fce00ddc75dc1154581e8df47e5f0..f6502be87c590ac3caa67a9b733c16dcd9b57cb1#r123573047

File: tests/generation/FileGlobs.py [hunk 10]
```diff
@@ -82,30 +102,12 @@ def test_get_project_files_gitignore(self):
...
 
         gitignore_dir_list = [os.getcwd(),
-                              os.path.join(os.getcwd(),"another_folder/"),
-                              os.path.join(os.getcwd(),"data","sample")]
+                              os.path.join(os.getcwd(), "another_folder"),
```
Comment (satwikkansal) [inline tests/generation/FileGlobs.py]: Same as https://github.com/coala/coala-quickstart/pull/110/files/749fc0d37f4fce00ddc75dc1154581e8df47e5f0..f6502be87c590ac3caa67a9b733c16dcd9b57cb1#r123573047

**Comment by Wert1996** [comment]
Hey @satwikkansal . Can you be a bit clearer on what changes you expect me to make from here? 
Thanks ;)

**Comment by satwikkansal** [comment]
Hey @Wert1996 , just remove styling mistake fixes in existing code from your PR. Example, https://github.com/coala/coala-quickstart/pull/110#discussion-diff-123573047R86. Hope it's clear now :)

**Comment by Wert1996** [comment]
@satwikkansal  The styling fixes were in my own added code, not in the original code. Should I still remove them? :)

**Comment by satwikkansal** [comment]
And please squash the three commits into one :)

Also rebase :sweat_smile:

**Comment by satwikkansal** [comment]
Thanks @Wert1996, this PR looks great to me. Let's wait for some maintainer to do a final review and merge :)

**Comment by Wert1996** [comment]
Thanks @satwikkansal :)

**Comment by Asnelchristian** [comment]
this is okay

**Comment by Asnelchristian** [comment]
just rebase and resolve the conflicts

**Comment by Wert1996** [comment]
@Asnelchristian messed up a bit but managed in the end. :)

File: coala_quickstart/generation/Utilities.py [hunk 11]
```diff
@@ -57,23 +57,26 @@ def parse_gitignore_line(line):
...
     """
     Generates a list of glob expressions equivalent to the
     contents of the user's project's ``.gitignore`` file.
 
     :param project_dir:
         The user's project directory.
+    :param gitignore_dir_list:
+        A list of paths of the directories in the project
```
Comment (Asnelchristian) [inline coala_quickstart/generation/Utilities.py]: How about:
```
List of paths to dicrectories in the project where .gitignore file exists
```

Comment (Asnelchristian) [inline coala_quickstart/generation/Utilities.py]: any way try formulating this better

**Comment by Asnelchristian** [comment]
almost there @Wert1996 :+1:, just one thing remaining

**Comment by satwikkansal** [comment]
ack a6c6162

**Comment by nkprince007** [comment]
ack https://github.com/coala/coala-quickstart/pull/110/commits/a6c6162ca9f4c6264c23bbb8271d1c5c0b8ba6ca

**Comment by satwikkansal** [comment]
@Wert1996 around? Sorry for keeping this PR on hold for too long (infact, very very long). Now that I have the merging rights, could you please rebase, push, and then ping me to merge.

**Comment by sils** [comment]
this has conflicts. Marking WIP

**Comment by ishanSrt** [comment]
PR used, fixed merge conflict and migrated to https://github.com/coala/coala-quickstart/pull/236 due to inactivity

**Comment by satwikkansal** [comment]
Closing this on account of https://github.com/coala/coala-quickstart/pull/236, author's commits are duly credited there.
